datameta/api/password.py

```diff
@@ -0,0 +1,88 @@
+# Copyright (c) 2021 Universität Tübingen, Germany
+# Authors: Leon Kuchenbecker <leon.kuchenbecker@uni-tuebingen.de>,
+#          Kersten Breuer <k.breuer@dkfz.de>
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+from pyramid.httpexceptions import HTTPNoContent, HTTPNotFound, HTTPForbidden, HTTPBadRequest
+
+from pyramid.view import view_config
+
+from .. import security
+from ..models import User, PasswordToken
+from uuid import UUID
+from datetime import datetime
+
+@view_config(
+    route_name="SetUserPassword",
+    renderer='json',
+    request_method="PUT",
+    openapi=True
+)
+def post(request):
+    db = request.dbsession
+
+    request_id = request.matchdict['id']
+    request_newPassword = request.openapi_validated.body["newPassword"]
+    request_credential = request.openapi_validated.body["passwordChangeCredential"]
+
+    # Case A: Password reset token
+    token = None
+    if request_id=='0':
+        token = db.query(PasswordToken).filter(PasswordToken.value==request_credential).one_or_none()
+        if token is not None and token.expires < datetime.now():
+            raise HTTPBadRequest(json=[{'message':'Token expired.'}])
```

Comment:
Maybe custom error in `errors.py` 410, as discussed

Reply:
Yup, can you merge that, then I'll rebase